src/aeso_ast_infer_types.erl

```diff
@@ -2556,55 +2694,94 @@ unify1(_Env, {uvar, A, R}, T, When) ->
             ets_insert(type_vars, {R, T}),
             true
     end;
-unify1(Env, T, {uvar, A, R}, When) ->
-    unify1(Env, {uvar, A, R}, T, When);
-unify1(_Env, {tvar, _, X}, {tvar, _, X}, _When) -> true; %% Rigid type variables
-unify1(Env, [A|B], [C|D], When) ->
-    unify(Env, A, C, When) andalso unify(Env, B, D, When);
-unify1(_Env, X, X, _When) ->
+unify1(Env, T, {uvar, A, R}, Variance, When) ->
+    unify1(Env, {uvar, A, R}, T, Variance, When);
+unify1(_Env, {tvar, _, X}, {tvar, _, X}, _Variance, _When) -> true; %% Rigid type variables
+unify1(Env, [A|B], [C|D], [V|Variances], When) ->
+    unify0(Env, A, C, V, When) andalso unify0(Env, B, D, Variances, When);
+unify1(Env, [A|B], [C|D], Variance, When) ->
+    unify0(Env, A, C, Variance, When) andalso unify0(Env, B, D, Variance, When);
+unify1(_Env, X, X, _Variance, _When) ->
     true;
-unify1(_Env, {id, _, Name}, {id, _, Name}, _When) ->
+unify1(_Env, {id, _, Name}, {id, _, Name}, _Variance, _When) ->
     true;
-unify1(_Env, {con, _, Name}, {con, _, Name}, _When) ->
-    true;
-unify1(_Env, {qid, _, Name}, {qid, _, Name}, _When) ->
+unify1(Env, A = {con, _, NameA}, B = {con, _, NameB}, Variance, When) ->
+    IsSubtype = case Variance of
```

Comment:
This could be a `is_subtype/4` that calls `is_subtype/3`, less indentation

Reply:
Done here cd3da44973ca25574079ae4f1c1a930c8828d6d6